action/sealedenvelope.go

```diff
@@ -7,19 +7,50 @@ import (
 	"github.com/iotexproject/iotex-proto/golang/iotextypes"
 	"github.com/pkg/errors"
 
+	"github.com/iotexproject/iotex-core/config"
 	"github.com/iotexproject/iotex-core/pkg/util/byteutil"
 )
 
 // SealedEnvelope is a signed action envelope.
 type SealedEnvelope struct {
 	Envelope
+	encoding      iotextypes.Encoding
+	externChainID uint32
+	srcPubkey     crypto.PublicKey
+	signature     []byte
+}
 
-	srcPubkey crypto.PublicKey
-	signature []byte
+// rawHash returns the raw hash of embedded Envelope (this is the hash to be signed)
+// an all-0 return value means the transaction is invalid
+func (sealed *SealedEnvelope) rawHash() hash.Hash256 {
+	if IsRLP(sealed.encoding) {
+		tx, err := actionToRLP(sealed.Action())
+		if err != nil {
+			return hash.ZeroHash256
+		}
+		h, err := rlpRawHash(tx, sealed.externChainID)
+		if err != nil {
+			return hash.ZeroHash256
+		}
+		return h
+	}
+	return sealed.Envelope.Hash()
 }
 
 // Hash returns the hash value of SealedEnvelope.
+// an all-0 return value means the transaction is invalid
 func (sealed *SealedEnvelope) Hash() hash.Hash256 {
+	if IsRLP(sealed.encoding) {
+		tx, err := actionToRLP(sealed.Action())
+		if err != nil {
```

Comment:
we should panic here, but let's leave it to next PR

Reply:
node receive tx thru API endpoint --> tx.LoadProto() --> actpool.Add(tx) --> call Verify(tx), if pass, tx is added to actpool, if fail, tx is rejected

if panic here, Verify() will panic, and node stop running
return all-0 hash will make Verify() fail gracefully, reject this tx, and continue running. i think this is better than panic